flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -426,6 +462,41 @@ public boolean isShutdown() {
         return shutdown;
     }
 
+    /**
+     * Reports whether a source operator is currently processing backlog.
+     *
+     * <p>If any source operator is processing backlog, the checkpoint interval would be decided by
+     * {@code execution.checkpointing.interval-during-backlog} instead of {@code
+     * execution.checkpointing.interval}.
+     *
+     * <p>If a source has not invoked this method, the source is considered to have
+     * isProcessingBacklog=false. If a source operator has invoked this method multiple times, the
+     * last reported value is used.
+     *
+     * @param operatorID the operator ID of the source operator.
+     * @param isProcessingBacklog whether the source operator is processing backlog.
+     */
+    public void setIsProcessingBacklog(OperatorID operatorID, boolean isProcessingBacklog) {
+        boolean wasBacklog = !backlogOperators.isEmpty();
+        if (isProcessingBacklog) {
+            backlogOperators.add(operatorID);
+        } else {
+            backlogOperators.remove(operatorID);
+        }
+
+        if (wasBacklog && backlogOperators.isEmpty()) {
+            baseInterval = checkpointInterval;
+        } else if (!wasBacklog && !backlogOperators.isEmpty()) {
+            baseInterval = checkpointIntervalDuringBacklog;
+        } else {
+            return;
+        }
+
+        if (baseInterval != Long.MAX_VALUE) {
+            scheduleTriggerWithDelay(getRandomInitDelay());
```

Comment:
The existing behavior introduce random initial delay only once, partially because it only needs to schedule periodic checkpoint once. Thus I don't think there is a suitable enough existing behavior to follow. Maybe we can discuss offline.